### ISSUE TITLE
[Health] Replace Stub Health Metrics System with Production Implementation

### ISSUE THREAD

**Issue by EffortlessSteven** [EffortlessMetrics/BitNet-rs#370]
# [Health] Replace Stub Health Metrics System with Production Implementation

## Problem Description

The BitNet.rs health monitoring system contains multiple stub implementations that return hardcoded placeholder values instead of performing real health checks and metrics collection. This affects production readiness and prevents accurate monitoring of system health, performance degradation detection, and operational observability.

## Environment

- **Crate**: `bitnet-server` (monitoring module)
- **Files Affected**:
  - `crates/bitnet-server/src/monitoring/health.rs`
  - `crates/bitnet-server/src/monitoring/metrics.rs`
  - `crates/bitnet-inference/src/gpu.rs`
  - `crates/bitnet-inference/src/generation/autoregressive.rs`
- **Features**: All (CPU and GPU)
- **Dependencies**: `sysinfo`, `metrics`, `tokio`, `chrono`

### 1. Hardcoded Health Check Values
In `health.rs`, several critical health checks return placeholder values:

```rust
// Line 196: Memory usage is hardcoded
let memory_usage_percent = 45.0; // Placeholder

// Lines 298-306: Health metrics are all zeros
async fn collect_health_metrics(&self) -> HealthMetrics {
    HealthMetrics {
        active_requests: 0,
        total_requests: 0,
        error_rate_percent: 0.0,
        avg_response_time_ms: 0.0,
        memory_usage_mb: 0.0,
        tokens_per_second: 0.0,
    }
}

// Line 312: CUDA version is hardcoded
fn get_cuda_version(&self) -> String {
    "12.3".to_string()
}
```

### 2. Stub Performance Metrics Collection
In `metrics.rs`, performance snapshots contain placeholder values:

```rust
// Lines 224-228: All performance metrics are hardcoded to 0.0
let snapshot = PerformanceSnapshot {
    timestamp: Instant::now(),
    tokens_per_second: 0.0, // Should be calculated from recent history
    memory_usage_mb: 0.0,   // Should be from actual memory usage
    active_requests: 0.0,   // Should be from the gauge
    error_rate: 0.0,        // Should be calculated from error counts
};
```

### 3. GPU Performance Metrics Placeholders
In `gpu.rs`, GPU utilization is hardcoded:

```rust
// GPU utilization placeholder
metrics.gpu_utilization = 0.85; // Placeholder
```

### 4. Generation Statistics Stubs
In `autoregressive.rs`, several `GenerationStats` fields are not tracked:

```rust
pub struct GenerationStats {
    // These fields are never properly updated:
    pub temperature_adjustments: usize,
    pub fallback_to_greedy: usize,
    pub batched_generations: usize,
    pub average_entropy: f64,
}
```

### Severity: High
- **Monitoring Blindness**: Unable to detect actual system health degradation
- **Performance Regression Detection**: Cannot identify performance drops or bottlenecks
- **Resource Management**: Cannot track actual memory/GPU usage for scaling decisions
- **Operational Issues**: Load balancers and orchestrators receive inaccurate health data

### Affected Users
- **DevOps/SRE Teams**: Cannot properly monitor production deployments
- **Development Teams**: Cannot identify performance regressions during development
- **Production Users**: System may appear healthy when experiencing issues

#### 1.1 Memory Health Monitoring
Replace hardcoded memory percentage with actual system metrics:

```rust
async fn check_memory_health(&self) -> ComponentHealth {
    let start = Instant::now();

    // Use actual system memory information
    let mut sys = System::new();
    sys.refresh_memory();

    let total_memory = sys.total_memory();
    let used_memory = sys.used_memory();
    let memory_usage_percent = if total_memory > 0 {
        (used_memory as f64 / total_memory as f64) * 100.0
    } else {
        0.0
    };

    let (status, message) = if memory_usage_percent > 90.0 {
        (HealthStatus::Unhealthy, format!("Critical memory usage: {:.1}%", memory_usage_percent))
    } else if memory_usage_percent > 80.0 {
        (HealthStatus::Degraded, format!("High memory usage: {:.1}%", memory_usage_percent))
    } else {
        (HealthStatus::Healthy, format!("Memory usage normal: {:.1}%", memory_usage_percent))
    };

    ComponentHealth {
        status,
        message,
        last_check: chrono::Utc::now().to_rfc3339(),
        response_time_ms: Some(start.elapsed().as_millis() as u64),
    }
}
```

#### 1.2 Model Health Validation
Replace placeholder model health with actual model accessibility checks:

```rust
async fn check_model_health(&self) -> ComponentHealth {
    let start = Instant::now();

    // Attempt to access model metadata or perform lightweight inference test
    let (status, message) = match self.validate_model_accessibility().await {
        Ok(model_info) => (HealthStatus::Healthy, format!("Model accessible: {}", model_info)),
        Err(e) => (HealthStatus::Unhealthy, format!("Model inaccessible: {}", e)),
    };

    ComponentHealth {
        status,
        message,
        last_check: chrono::Utc::now().to_rfc3339(),
        response_time_ms: Some(start.elapsed().as_millis() as u64),
    }
}

async fn validate_model_accessibility(&self) -> Result<String, String> {
    // Perform actual model validation
    // This could involve checking if model files exist, are readable,
    // and can perform a minimal inference test
}
```

#### 1.3 GPU Health Monitoring
Implement real GPU health checks using CUDA APIs:

```rust
#[cfg(feature = "cuda")]
async fn check_gpu_health(&self) -> ComponentHealth {
    let start = Instant::now();

    let (status, message) = match self.get_real_gpu_status().await {
        Ok(gpu_info) => {
            if gpu_info.memory_usage_percent > 95.0 {
                (HealthStatus::Degraded, format!("GPU memory high: {:.1}%", gpu_info.memory_usage_percent))
            } else if gpu_info.temperature > 85.0 {
                (HealthStatus::Degraded, format!("GPU temperature high: {:.1}°C", gpu_info.temperature))
            } else {
                (HealthStatus::Healthy, format!("GPU healthy: {}MB free", gpu_info.free_memory_mb))
            }
        },
        Err(e) => (HealthStatus::Unhealthy, format!("GPU unavailable: {}", e)),
    };

    ComponentHealth {
        status,
        message,
        last_check: chrono::Utc::now().to_rfc3339(),
        response_time_ms: Some(start.elapsed().as_millis() as u64),
    }
}

#[cfg(feature = "cuda")]
async fn get_real_gpu_status(&self) -> Result<GpuStatus, String> {
    // Use actual CUDA APIs to query GPU status
    // This should check: memory usage, temperature, utilization, error state
}
```

#### 2.1 Performance Snapshot Collection
Replace hardcoded zeros with actual metric values:

```rust
async fn collect_system_metrics(&self) -> Result<()> {
    // ... existing code for uptime and memory ...

    // Collect real performance metrics
    let active_requests = self.get_active_request_count();
    let tokens_per_second = self.calculate_recent_tokens_per_second().await;
    let error_rate = self.calculate_recent_error_rate().await;
    let memory_usage_mb = self.get_current_memory_usage_mb().await;

    let snapshot = PerformanceSnapshot {
        timestamp: Instant::now(),
        tokens_per_second,
        memory_usage_mb,
        active_requests,
        error_rate,
    };

    // ... rest of the function ...
}

async fn calculate_recent_tokens_per_second(&self) -> f64 {
    // Calculate actual tokens/second from recent inference operations
}

async fn calculate_recent_error_rate(&self) -> f64 {
    // Calculate error rate from recent request history
}
```

#### 2.2 Health Metrics Collection
Replace placeholder `collect_health_metrics` with real implementation:

```rust
async fn collect_health_metrics(&self) -> HealthMetrics {
    let active_requests = self.metrics.inference.requests_active.value() as u64;
    let total_requests = self.metrics.inference.requests_total.value() as u64;

    // Calculate error rate from recent history
    let error_rate_percent = self.calculate_error_rate_percentage().await;

    // Get average response time from histogram
    let avg_response_time_ms = self.get_average_response_time_ms().await;

    // Get current memory usage
    let memory_usage_mb = self.get_current_memory_usage_mb().await;

    // Get recent tokens per second
    let tokens_per_second = self.get_recent_tokens_per_second().await;

    HealthMetrics {
        active_requests,
        total_requests,
        error_rate_percent,
        avg_response_time_ms,
        memory_usage_mb,
        tokens_per_second,
    }
}
```

#### 3.1 Real GPU Utilization Monitoring
Replace hardcoded GPU utilization with actual CUDA queries:

```rust
// In gpu.rs
#[cfg(feature = "cuda")]
pub fn get_gpu_utilization(&self) -> Result<f64, String> {
    // Query actual GPU utilization using CUDA API or nvidia-ml-py equivalent
    use nvidia_ml_rs::*;

    match nvml::init() {
        Ok(nvml) => {
            match nvml.device_by_index(self.device_id) {
                Ok(device) => {
                    match device.utilization_rates() {
                        Ok(utilization) => Ok(utilization.gpu as f64 / 100.0),
                        Err(e) => Err(format!("Failed to get GPU utilization: {}", e)),
                    }
                },
                Err(e) => Err(format!("Failed to get GPU device: {}", e)),
            }
        },
        Err(e) => Err(format!("Failed to initialize NVML: {}", e)),
    }
}

// In generate_tokens_gpu:
metrics.gpu_utilization = self.get_gpu_utilization().unwrap_or(0.0);
```

#### 4.1 Implement Generation Metrics Tracking
Add real tracking for generation statistics:

```rust
impl AutoregressiveGenerator {
    pub fn new(config: GenerationConfig) -> Self {
        Self {
            // ... existing fields ...
            stats_tracker: GenerationStatsTracker::new(),
        }
    }

    async fn sample_next_token_with_prob(&mut self, logits: &[f32]) -> Result<u32> {
        // Track temperature adjustments
        if self.adjusted_temperature_this_step {
            self.stats_tracker.temperature_adjustments += 1;
        }

        // ... existing sampling logic ...
    }

    async fn consider_sampling_fallback(&mut self) -> Result<()> {
        match self.performance_mode {
            PerformanceMode::Conservative => {
                // Track fallback to greedy
                self.stats_tracker.fallback_to_greedy += 1;
                // ... existing logic ...
            }
        }
    }

    async fn generate_token_batched(&mut self, batch: &[InferenceRequest]) -> Result<()> {
        self.stats_tracker.batched_generations += 1;
        // ... existing logic ...
    }

    pub fn get_stats(&self) -> GenerationStats {
        GenerationStats {
            // ... existing stats ...
            temperature_adjustments: self.stats_tracker.temperature_adjustments,
            fallback_to_greedy: self.stats_tracker.fallback_to_greedy,
            batched_generations: self.stats_tracker.batched_generations,
            average_entropy: self.stats_tracker.calculate_average_entropy(),
        }
    }
}

struct GenerationStatsTracker {
    temperature_adjustments: usize,
    fallback_to_greedy: usize,
    batched_generations: usize,
    entropy_history: VecDeque<f64>,
}

impl GenerationStatsTracker {
    fn calculate_average_entropy(&self) -> f64 {
        if self.entropy_history.is_empty() {
            0.0
        } else {
            self.entropy_history.iter().sum::<f64>() / self.entropy_history.len() as f64
        }
    }
}
```

### Task 1: Health Check Foundation (2-3 days)
- [ ] Implement real memory usage monitoring in `check_memory_health`
- [ ] Add model accessibility validation in `check_model_health`
- [ ] Create inference engine responsiveness test in `check_inference_engine_health`
- [ ] Add comprehensive error handling and timeouts

### Task 2: GPU Health Monitoring (2-3 days)
- [ ] Research and implement CUDA health query APIs
- [ ] Add GPU memory usage monitoring
- [ ] Implement GPU temperature monitoring
- [ ] Add GPU utilization tracking
- [ ] Handle GPU unavailable/error states gracefully

### Task 3: Metrics Collection Infrastructure (3-4 days)
- [ ] Implement real-time metrics aggregation system
- [ ] Add sliding window calculations for rates and averages
- [ ] Create performance history management
- [ ] Implement error rate calculation
- [ ] Add memory usage tracking integration

### Task 4: Generation Statistics Tracking (2-3 days)
- [ ] Add generation statistics tracker to AutoregressiveGenerator
- [ ] Implement temperature adjustment tracking
- [ ] Add greedy fallback detection
- [ ] Implement batch generation counting
- [ ] Add entropy calculation for quality metrics

### Task 5: Integration and Testing (2-3 days)
- [ ] Integrate all real implementations
- [ ] Add comprehensive unit tests for health checks
- [ ] Add integration tests for metrics collection
- [ ] Performance test metrics collection overhead
- [ ] Add error injection tests for failure scenarios

### Task 6: Configuration and Documentation (1-2 days)
- [ ] Add configuration options for health check thresholds
- [ ] Document new health check behaviors
- [ ] Update monitoring setup documentation
- [ ] Add troubleshooting guide for health check failures

### Unit Tests
```rust
#[cfg(test)]
mod tests {
    use super::*;

    #[tokio::test]
    async fn test_memory_health_check_accuracy() {
        let health_checker = HealthChecker::new(Arc::new(mock_metrics()));
        let result = health_checker.check_memory_health().await;

        // Verify actual memory usage is reported
        assert_ne!(result.message, "Memory usage normal: 45.0%");
        assert!(result.message.contains("Memory usage"));
        assert!(result.response_time_ms.is_some());
    }

    #[tokio::test]
    async fn test_metrics_collection_real_values() {
        let health_checker = HealthChecker::new(Arc::new(mock_metrics()));
        let metrics = health_checker.collect_health_metrics().await;

        // Verify metrics are not all zeros (when system is active)
        // Note: In a real system with activity, these shouldn't all be zero
        assert!(metrics.memory_usage_mb >= 0.0);
    }

    #[cfg(feature = "cuda")]
    #[tokio::test]
    async fn test_gpu_health_real_status() {
        let health_checker = HealthChecker::new(Arc::new(mock_metrics()));
        let result = health_checker.check_gpu_health().await;

        // Should either succeed with real GPU info or fail with meaningful error
        match result.status {
            HealthStatus::Healthy => assert!(result.message.contains("GPU")),
            HealthStatus::Unhealthy => assert!(result.message.contains("unavailable")),
            HealthStatus::Degraded => assert!(result.message.contains("high")),
        }
    }
}
```

### Integration Tests
```rust
#[tokio::test]
async fn test_health_endpoint_real_metrics() {
    let app = test_app().await;
    let response = app.get("/health").await;

    assert_eq!(response.status(), 200);
    let health: HealthResponse = response.json().await;

    // Verify real values are returned
    assert_ne!(health.metrics.memory_usage_mb, 0.0);
    assert!(health.uptime_seconds > 0);
    assert_ne!(health.components.len(), 0);
}
```

### Performance Tests
```rust
#[tokio::test]
async fn test_health_check_performance() {
    let health_checker = HealthChecker::new(Arc::new(mock_metrics()));

    let start = Instant::now();
    let _result = health_checker.check_health().await;
    let duration = start.elapsed();

    // Health checks should complete quickly
    assert!(duration < Duration::from_millis(100));
}
```

### Functional Requirements
- [ ] Health checks return actual system state, not hardcoded values
- [ ] Memory usage reflects real system memory consumption
- [ ] GPU health checks use actual CUDA APIs when available
- [ ] Metrics collection aggregates real performance data
- [ ] Generation statistics track actual inference behavior
- [ ] Error rates calculated from real request history

### Performance Requirements
- [ ] Health check endpoint responds in <100ms under normal load
- [ ] Metrics collection adds <1% overhead to inference requests
- [ ] GPU health checks complete in <50ms
- [ ] Memory usage tracking accuracy within 5% of system tools

### Reliability Requirements
- [ ] Health checks handle GPU unavailability gracefully
- [ ] System continues operating if individual health checks fail
- [ ] Metrics collection resilient to temporary metric calculation failures
- [ ] Health status degrades gracefully under resource pressure

### Monitoring Requirements
- [ ] All health components report meaningful status messages
- [ ] Performance regression detection works with real metrics
- [ ] Error rate calculation identifies actual system issues
- [ ] GPU utilization tracking matches nvidia-smi output (within 5%)

### External Dependencies
- Add to Cargo.toml:
```toml
[dependencies]
nvidia-ml-rs = { version = "0.4", optional = true }
sysinfo = "0.30"

[features]
cuda = ["nvidia-ml-rs"]
```

### Internal Dependencies
- Integration with existing `MetricsCollector`
- Coordination with `bitnet-inference` crate metrics
- Configuration system integration for thresholds

### High Risk Items
1. **GPU API Integration**: CUDA/NVML APIs may not be available in all environments
2. **Performance Impact**: Real metrics collection may add latency
3. **Platform Compatibility**: System metrics may behave differently across platforms

### Mitigation Strategies
- Graceful degradation when GPU APIs unavailable
- Async metrics collection to minimize request impact
- Platform-specific metric collection implementations
- Comprehensive fallback mechanisms for failed health checks

## Related Issues

This issue addresses stub implementations identified in the production readiness audit. Related issues include:

- GPU Memory Management System implementation
- Backend Functions completion
- Validation System implementation
- Configuration System improvements

## Labels

- `production-readiness`
- `monitoring`
- `health-checks`
- `metrics`
- `gpu`
- `priority-high`
- `multi-component`

## Estimated Effort

**Total: 12-18 days**
- Development: 10-14 days
- Testing: 2-3 days
- Documentation: 1-2 days

This represents a substantial but necessary investment to enable reliable production monitoring and operational observability.
